File: 91-decode-ways/leochang/91.cpp
class Solution {
public:
    int numDecodings(string s) {
        if (0 == s.size() || s[0] == '0') return 0;
        
        vector<int> dp(s.size()+1, 0);
        dp[0] = 1;
        dp[1] = (s[1] != '0') ? 1 : 0;
        
        for (int i = 2; i <= s.size(); i++) {
            if (i < s.size() && s[i] == '0') {
                dp[i] = 0;
            } else if (s[i-2] == '1' || (s[i-2] == '2' && s[i-1] <= '6')) {
                dp[i] = dp[i-1] + dp[i-2];
            } else {
                dp[i] = dp[i-1];
            }
        }
        return dp[s.size()];
    }
};

File: 91-decode-ways/ken/91.py
class Solution(object):
    def __init__(self):
        self.dp = {}

    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """

        if s in self.dp:
            return self.dp[s]
        if len(s) == 0 or s[0] == '0':
            return 0
        if len(s) == 1:
            return 1

        n = 0
        if len(s[:2]) == 2 and s[:2] > '0' and s[:2] <= '26':
            if (len(s[2:]) == 0):
                n += 1
            else:
                n += self.numDecodings(s[2:])

        n += self.numDecodings(s[1:])

        if s not in self.dp:
            self.dp[s] = n

        return n
